sdks/agent-sdk/src/core/Agent.ts

```diff
@@ -191,58 +190,68 @@ export class Agent<ContentTypes = unknown> extends EventEmitter<
     return this;
   }
 
+  async #handleStreamError(
+    error: unknown,
+    options?: StreamAllMessagesOptions<ContentTypes>,
+  ) {
+    this.#messageStream = undefined;
+    const recovered = await this.#runErrorChain(error, {
+      client: this.#client,
+    });
+    if (recovered) {
+      queueMicrotask(() => this.start(options));
+    }
+  }
+
   async start(options?: StreamAllMessagesOptions<ContentTypes>) {
-    if (this.#isListening) {
+    if (this.#messageStream) {
       return;
     }
 
     try {
-      this.#isListening = true;
-      void this.emit("start");
-
-      this.#messageStream =
-        await this.#client.conversations.streamAllMessages(options);
-      for await (const message of this.#messageStream) {
-        // The "stop()" method sets "isListening"
-        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-        if (!this.#isListening) break;
-        try {
-          switch (true) {
-            case filter.isRemoteAttachment(message):
-              await this.#processMessage(message, "attachment");
-              break;
-            case filter.isReaction(message):
-              await this.#processMessage(message, "reaction");
-              break;
-            case filter.isReply(message):
-              await this.#processMessage(message, "reply");
-              break;
-            case filter.isText(message):
-              await this.#processMessage(message, "text");
-              break;
-            default:
-              await this.#processMessage(message);
-              break;
+      this.#messageStream = await this.#client.conversations.streamAllMessages({
+        ...options,
```

Comment:
since we're overwriting these callbacks, we should also run the ones passed in as options.